src/goto-programs/destructor_tree.cpp

```diff
@@ -0,0 +1,87 @@
+// Copyright 2016-2018 Diffblue Limited.
+
+#include "destructor_tree.h"
+
+void destructor_treet::add(const codet &destructor)
+{
+  // Attach the new node to our most recent node or rewind target.
+  auto previous_node = get_current_node();
+  auto new_node = destruction_graph.add_node(destructor);
+  if(rewind_target)
+  {
+    auto rewind_target_index = *rewind_target;
+    rewind_target = {};
+    destruction_graph.add_edge(rewind_target_index, new_node);
+  }
+  else
+  {
+    destruction_graph.add_edge(previous_node, new_node);
+  }
+}
+
+optionalt<codet> &destructor_treet::get_destructor(node_indext index)
+{
+  PRECONDITION(index < destruction_graph.size());
+  return destruction_graph[index].destructor_value;
+}
+
+const ancestry_resultt destructor_treet::get_nearest_common_ancestor_info(
+  node_indext left_index,
+  node_indext right_index)
+{
+  if(destruction_graph.empty())
+    return ancestry_resultt(0);
+
+  node_indext left_unique_count = 0, right_unique_count = 0;
```

Comment:
You might have taken my suggestion of replacing `std::size_t` one step too far: those counters should IMHO remain `std::size_t`, they aren't indices.